Fail to compile after  [419877cf] commit (Added -Wconversion support for C++.)

I see there is fixes for correct type cast.

Previously within pack.hpp was declaraion:

    void append_buffer(const char* buf, size_t len)
        { m_stream.write(buf, len); }

Now there is couple of compile time overloads to respect valid type cast when **len** parameter is not size_t.   But, what if method **::write** has overloads ? In this case **&Stream::write** will not be compiled (because compiler has ambiguity). 

I have class Writer with 'primary' **write** method

virtual void write(const std::byte*, size_t size) = 0;

and second 'utility' method:

	void write(const char* bytes, size_t size)
	{
		write(reinterpret_cast<const std::byte*>(bytes), size);
	}

Second method introduced especially to using with msgpack-c, but for **first** parameter compatibility. As temporary solution i remove second 'utility' method from my Writer class and 'fix' pack.hpp as

    void append_buffer(const char* buf, size_t len)
    {
        append_buffer(m_stream, &Stream::write, buf, len);
    }

    template<typename S, typename R, typename Ptr, typename Size>
    static void append_buffer(S& s, R (S::*) (Ptr, Size), const char* buf, size_t len)
    {
      s.write(reinterpret_cast<Ptr>(buf), static_cast<Size>(len));
    }

> Now there is couple of compile time overloads to respect valid type cast when len parameter is not size_t. But, what if method ::write has overloads ? In this case &Stream::write will not be compiled (because compiler has ambiguity).

Is your error the same as follows? 

```cpp
#include <cstdlib>
#include <utility>
#include <ostream>

template <typename Stream>
struct packer {
    packer(Stream& s):m_stream(s) {}
    void append_buffer(const char* buf, size_t len)
    {
        append_buffer(m_stream, &Stream::write, buf, len);
    }

    template <typename S, typename Write>
    typename std::enable_if<
        std::is_same<
            std::ostream& (std::ostream::*)(const char*, std::streamsize),
            Write
        >::value
    >::type
    append_buffer(S& s, Write, const char* buf, size_t len)
    {
        s.write(buf, static_cast<std::streamsize>(len));
    }

    template <typename S, typename Write>
    typename std::enable_if<
        !std::is_same<
            std::ostream& (std::ostream::*)(const char*, std::streamsize),
            Write
        >::value
    >::type
    append_buffer(S& s, Write, const char* buf, size_t len)
    {
        s.write(buf, len);
    }

    Stream& m_stream;
};

struct user_stream {
    void write(const char* bytes, size_t size) {}
    void write(const std::byte* bytes, size_t size) {}
};


int main() {
    user_stream us;
    packer<user_stream> p(us);
    char buf[] = "123";

    p.append_buffer(buf, sizeof(buf));
}
```

```
issue777.cpp:10:9: error: no matching member function for call to 'append_buffer'
        append_buffer(m_stream, &Stream::write, buf, len);
        ^~~~~~~~~~~~~
issue777.cpp:51:7: note: in instantiation of member function 'packer<user_stream>::append_buffer' requested here
    p.append_buffer(buf, sizeof(buf));
      ^
issue777.cpp:20:5: note: candidate template ignored: couldn't infer template argument 'Write'
    append_buffer(S& s, Write, const char* buf, size_t len)
    ^
issue777.cpp:32:5: note: candidate template ignored: couldn't infer template argument 'Write'
    append_buffer(S& s, Write, const char* buf, size_t len)
    ^
issue777.cpp:8:10: note: candidate function not viable: requires 2 arguments, but 4 were provided
    void append_buffer(const char* buf, size_t len)
         ^
1 error generated.
```

Hmm.... I wrote a minimal code to check overloaded functions.

```cpp
#include <cstddef>

struct user_stream3 {
    void write(const char* bytes, size_t size) {}
    void write(const std::byte* bytes, size_t size) {}
};


template <typename Write>
void foo(Write) {}

int main() {
    foo(&user_stream3::write);
}
```

And got

```
clang++ -std=c++17 -Wconversion issue777_4.cpp 
issue777_4.cpp:13:5: error: no matching function for call to 'foo'
    foo(&user_stream3::write);
    ^~~
issue777_4.cpp:10:6: note: candidate template ignored: couldn't infer template argument 'Write'
void foo(Write) {}
     ^
1 error generated.

Compilation exited abnormally with code 1 at Wed May  1 08:56:58
```

In order to get a specific overloaded function, Function return type and parameter types are required. However, it is unpredictable. I think that there is no solution.

So, I revert the code and add pragma to avoid warnings on this place.

--- a/include/msgpack/v1/pack.hpp
+++ b/include/msgpack/v1/pack.hpp
@@ -620,31 +620,13 @@ private:
 
     void append_buffer(const char* buf, size_t len)
     {
-        append_buffer(m_stream, &Stream::write, buf, len);
+        append_buffer(&Stream::write, buf, len);
     }
 
-    template <typename S, typename Write>
-    typename enable_if<
-        is_same<
-            std::ostream& (std::ostream::*)(const char*, std::streamsize),
-            Write
-        >::value
-    >::type
-    append_buffer(S& s, Write, const char* buf, size_t len)
+    template <typename Ret, typename Cls, typename SizeType>
+    void append_buffer(Ret (Cls::*)(const char*, SizeType), const char* buf, size_t len)
     {
-        s.write(buf, static_cast<std::streamsize>(len));
-    }
-
-    template <typename S, typename Write>
-    typename enable_if<
-        !is_same<
-            std::ostream& (std::ostream::*)(const char*, std::streamsize),
-            Write
-        >::value
-    >::type
-    append_buffer(S& s, Write, const char* buf, size_t len)
-    {
-        s.write(buf, len);
+        m_stream.write(buf, static_cast<SizeType>(len));
     }
 
 private:

--- a/include/msgpack/v1/pack.hpp
+++ b/include/msgpack/v1/pack.hpp
@@ -623,29 +623,11 @@ private:
         append_buffer(&Stream::write, buf, len);
     }
 
-#if defined(MSGPACK_USE_CPP03)
-
     template <typename Ret, typename Cls, typename SizeType>
     void append_buffer(Ret (Cls::*)(const char*, SizeType), const char* buf, size_t len)
     {
         m_stream.write(buf, static_cast<SizeType>(len));
     }
-
-#else  // defined(MSGPACK_USE_CPP03)
-
-    template <typename Ret, typename Cls,  typename SizeType>
-    void append_buffer(Ret (Cls::*)(const char*, SizeType) noexcept(false), const char* buf, size_t len) noexcept(false)
-    {
-        m_stream.write(buf, static_cast<SizeType>(len));
-    }
-
-    template <typename Ret, typename Cls,  typename SizeType>
-    void append_buffer(Ret (Cls::*)(const char*, SizeType) noexcept(true), const char* buf, size_t len) noexcept(true)
-    {
-        m_stream.write(buf, static_cast<SizeType>(len));
-    }
-
-#endif // defined(MSGPACK_USE_CPP03)
 
 private:
     Stream& m_stream;

--- a/include/msgpack/v1/pack.hpp
+++ b/include/msgpack/v1/pack.hpp
@@ -631,7 +631,7 @@ private:
 
 #if !defined(MSGPACK_USE_CPP03)
     template <typename Ret, typename Cls,  typename SizeType>
-    void append_buffer(Ret (Cls::*)(const char*, SizeType) noexcept(true), const char* buf, size_t len)
+    void append_buffer(Ret (Cls::*)(const char*, SizeType) noexcept, const char* buf, size_t len) noexcept
     {
         m_stream.write(buf, static_cast<SizeType>(len));
     }

--- a/include/msgpack/v1/pack.hpp
+++ b/include/msgpack/v1/pack.hpp
@@ -623,19 +623,29 @@ private:
         append_buffer(&Stream::write, buf, len);
     }
 
+#if defined(MSGPACK_USE_CPP03)
+
     template <typename Ret, typename Cls, typename SizeType>
     void append_buffer(Ret (Cls::*)(const char*, SizeType), const char* buf, size_t len)
     {
         m_stream.write(buf, static_cast<SizeType>(len));
     }
 
-#if !defined(MSGPACK_USE_CPP03)
+#else  // defined(MSGPACK_USE_CPP03)
+
     template <typename Ret, typename Cls,  typename SizeType>
-    void append_buffer(Ret (Cls::*)(const char*, SizeType) noexcept, const char* buf, size_t len) noexcept
+    void append_buffer(Ret (Cls::*)(const char*, SizeType) noexcept(false), const char* buf, size_t len) noexcept(false)
     {
         m_stream.write(buf, static_cast<SizeType>(len));
     }
-#endif // !defined(MSGPACK_USE_CPP03)
+
+    template <typename Ret, typename Cls,  typename SizeType>
+    void append_buffer(Ret (Cls::*)(const char*, SizeType) noexcept(true), const char* buf, size_t len) noexcept(true)
+    {
+        m_stream.write(buf, static_cast<SizeType>(len));
+    }
+
+#endif // defined(MSGPACK_USE_CPP03)
 
 private:
     Stream& m_stream;

--- a/include/msgpack/v1/pack.hpp
+++ b/include/msgpack/v1/pack.hpp
@@ -629,6 +629,14 @@ private:
         m_stream.write(buf, static_cast<SizeType>(len));
     }
 
+#if !defined(MSGPACK_USE_CPP03)
+    template <typename Ret, typename Cls,  typename SizeType>
+    void append_buffer(Ret (Cls::*)(const char*, SizeType) noexcept(true), const char* buf, size_t len)
+    {
+        m_stream.write(buf, static_cast<SizeType>(len));
+    }
+#endif // !defined(MSGPACK_USE_CPP03)
+
 private:
     Stream& m_stream;
 

